src/uu/mknod/src/mknod.rs

```diff
@@ -31,168 +51,162 @@ fn makedev(maj: u64, min: u64) -> dev_t {
-                return 1;
-            }
+        // set umask back to original value
+        if set_umask {
+            libc::umask(last_umask);
         }
-        unsafe {
-            last_umask = libc::umask(0);
+
+        if errno == -1 {
+            let c_str = CString::new(NAME).expect("Failed to convert to CString");
+            // shows the error from the mknod syscall
+            libc::perror(c_str.as_ptr());
         }
+        errno
     }
+}
 
-    let mut ret = 0i32;
-    match matches.free.len() {
-        0 => show_usage_error!("missing operand"),
-        1 => show_usage_error!("missing operand after ‘{}’", matches.free[0]),
-        _ => {
-            let args = &matches.free;
-            let c_str = CString::new(args[0].as_str()).expect("Failed to convert to CString");
-
-            // Only check the first character, to allow mnemonic usage like
-            // 'mknod /dev/rst0 character 18 0'.
-            let ch = args[1]
-                .chars()
-                .next()
-                .expect("Failed to get the first char");
-
-            if ch == 'p' {
-                if args.len() > 2 {
-                    show_info!("{}: extra operand ‘{}’", NAME, args[2]);
-                    if args.len() == 4 {
-                        eprintln!("Fifos do not have major and minor device numbers.");
-                    }
-                    eprintln!("Try '{} --help' for more information.", NAME);
-                    return 1;
-                }
+pub fn uumain(args: impl uucore::Args) -> i32 {
+    // Linux-specific options, not implemented
+    // opts.optflag("Z", "", "set the SELinux security context to default type");
+    // opts.optopt("", "context", "like -Z, or if CTX is specified then set the SELinux or SMACK security context to CTX");
 
-                ret = _makenod(c_str, S_IFIFO | newmode, 0);
-            } else {
-                if args.len() < 4 {
-                    show_info!("missing operand after ‘{}’", args[args.len() - 1]);
-                    if args.len() == 2 {
-                        eprintln!("Special files require major and minor device numbers.");
-                    }
-                    eprintln!("Try '{} --help' for more information.", NAME);
-                    return 1;
-                } else if args.len() > 4 {
-                    show_usage_error!("extra operand ‘{}’", args[4]);
-                    return 1;
-                } else if !"bcu".contains(ch) {
-                    show_usage_error!("invalid device type ‘{}’", args[1]);
-                    return 1;
-                }
+    let matches = App::new(executable!())
+        .version(VERSION)
+        .help(USAGE)
+        .arg(
+            Arg::with_name("mode")
+                .short("m")
+                .long("mode")
+                .value_name("MODE")
+                .help("set file permission bits to MODE, not a=rw - umask"),
+        )
+        .arg(
+            Arg::with_name("name")
```

Comment:
Do we need adding new flags to mknod? I think this hurts compatability with the existing implementation, you can do something like this (then it becomes pretty much `getopts::Matches::free`):
https://github.com/uutils/coreutils/blob/c224b955dab80a55ff10937316331d8d21d9ebde/src/uu/mkfifo/src/mkfifo.rs#L49

Reply:
From my testing, it doesn't look like `name` is added as a flag in this case. What I tested:
- removed hard-coded help message and let clap generate it --> no `--name` flag shown
- invoked `mknod --name testfile p` --> `Found argument '--name' which wasn't expected,`

Perhaps it is because I also use `index(1)`, but in any case: clap seems to correctly see this as a positional argument.